package dmlite;

message SerialExtendedStat {
  optional SerialStat stat = 1;
  optional int32 parent = 2;
  optional int32 type = 3;
  optional string status = 4;
  optional string name = 5;
  optional string guid = 6;
  optional string csumtype = 7;
  optional string csumvalue = 8;
  optional string acl = 9;
  optional SerialExtendedAttributeList xattrlist = 10;
}

message SerialStat {
  optional int32 st_dev = 1;
  optional int32 st_ino = 2;
  optional int32 st_mode = 3;
  optional int32 st_nlink = 4;
  optional int32 st_uid = 5;
  optional int32 st_gid = 6;
  optional int32 st_rdev = 7;
  optional int64 st_size = 8;
  optional int64 st_access_time = 9;
  optional int64 st_modified_time = 10;
  optional int64 st_change_time = 11;
  optional int32 st_blksize = 12;
  optional int32 st_blocks = 13;
}

message SerialSymLink {
  optional int32 fileid = 1;
  optional string link = 2;
}

message SerialComment {
  optional string comment = 1;
}

enum CacheState {
  MISSING = 0;
  INVALID = 1;
  VALID = 2;
}

message SerialKeyList {
  required CacheState state = 2 [default = MISSING];
  repeated SerialKey key = 1;
}

message SerialKey {
  required string key = 1;
}

message SerialReplicaList {
  repeated SerialReplica replica = 1;
}

message SerialReplica {
  optional int64 replicaid = 1;
  optional int64 fileid = 2;
  optional int64 nbaccesses = 3;
  optional int64 atime = 4;
  optional int64 ptime = 5;
  optional int64 ltime = 6;
  optional string status = 7;
  optional string type = 8;
  optional string setname = 9;
  optional string pool = 10;
  optional string server = 11;
  optional string filesystem = 12;
  optional string url = 13;
  optional SerialExtendedAttributeList xattrlist = 14;

}

message SerialPoolList {
  repeated SerialPool pool = 1;
}

message SerialPool {
  optional string name = 1;
  optional string type = 2;
  //optional int64 defsize = 3;
  //optional int64 gc_start_thresh = 4;
  //optional int64 gc_stop_thresh = 5;
  //optional int64 def_lifetime = 6;
  //optional int64 defpintime = 7;
  //optional int64 max_lifetime = 8;
  //optional int64 maxpintime = 9;
  //optional string fss_policy = 10;
  //optional string gc_policy = 11;
  //optional string mig_policy = 12;
  //optional string rs_policy = 13;
  //optional SerialGroups groups = 14;
  //optional string ret_policy = 15;
  //optional string s_type = 16;
  //optional string meta = 17;

}

message SerialUrl {
  optional string scheme = 1;
  optional string domain = 2;
  optional int64 port = 3;
  optional string path = 4;
  optional string token = 5;
}

message SerialChunk {
  optional int64 offset = 1;
  optional int64 size = 2;
  optional SerialUrl url = 3;
}

message SerialExtendedAttributeList {
  repeated SerialExtendedAttribute xattr = 1;
}

message SerialExtendedAttribute {
  optional string name = 1;
  optional string value = 2;
}
